src/stirling/source_connectors/socket_tracer/protocols/mongodb/stitcher.cc

```diff
@@ -0,0 +1,198 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/stitcher.h"
+
+#include <string>
+#include <utility>
+#include <variant>
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/strings/str_replace.h>
+#include "src/common/base/base.h"
+
+#include "src/stirling/source_connectors/socket_tracer/protocols/common/interface.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+void FlattenSections(mongodb::Frame* frame) {
+  // Flatten the vector of sections containing vector of documents into a single string.
+  for (const auto& section : frame->sections) {
+    for (const auto& doc : section.documents) {
+      frame->frame_body.append(doc).append(" ");
+    }
+  }
+  frame->sections.clear();
+}
+
+RecordsWithErrorCount<mongodb::Record> StitchFrames(
+    absl::flat_hash_map<mongodb::stream_id_t, std::deque<mongodb::Frame>>* reqs,
+    absl::flat_hash_map<mongodb::stream_id_t, std::deque<mongodb::Frame>>* resps, State* state) {
+  std::vector<mongodb::Record> records;
+  int error_count = 0;
+
+  for (auto stream_id_it = state->stream_order.begin(); stream_id_it != state->stream_order.end();
+       stream_id_it++) {
+    auto& stream_id_pair = *stream_id_it;
+    auto stream_id = stream_id_pair.first;
+
+    // Find the stream ID's response deque.
+    auto resp_it = resps->find(stream_id);
+    if (resp_it == resps->end()) {
+      VLOG(1) << absl::Substitute("Did not find a response deque with the stream ID: $0",
+                                  stream_id);
+      continue;
+    }
+
+    // Response deque for the stream ID.
+    auto& resp_deque = resp_it->second;
+
+    // Find the stream ID's request deque.
+    auto req_it = reqs->find(stream_id);
+    if (req_it == reqs->end()) {
+      VLOG(1) << absl::Substitute("Did not find a request deque with the stream ID: $0", stream_id);
+      error_count++;
+      continue;
+    }
+
+    // Request deque for the stream ID.
+    auto& req_deque = req_it->second;
+
+    // Track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = 0;
+
+    // Loop until the first frame in the response deque finds a match with the oldest request that
+    // occured just before the response.
+    for (const auto& [idx, resp_frame] : Enumerate(resp_deque)) {
+      if (resp_frame.consumed) {
+        continue;
+      }
+
+      latest_resp_ts = resp_frame.timestamp_ns;
+
+      auto curr_resp = resp_frame;
+
+      // Find and insert all of the moreToCome frame(s) section data to the head response frame.
+      while (curr_resp.more_to_come) {
```

Comment:
I think putting the more to come logic in a dedicated function would help with the readability of this if it's possible.